### PR TITLE
server: Get vm network/disk statistics and update database per host

### PR DESCRIPTION
**Pull request by ustcweizhou**
### Description

This PR fixes an issue when collect vm network or disk statistics from hosts.

If cloudstack fails to get vm statistics from a host, the whole process will be terminated, and it will not get vm statistics from remaining hosts.

With this pr, the host with problem will be ignored, and cloudstack will continue on remaining hosts.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [X] Minor
- [ ] Trivial

### CODE REVIEW

File: server/src/main/java/com/cloud/server/StatsCollector.java [hunk 1]
```diff
@@ -793,10 +792,10 @@ public void doInTransactionWithoutResult(TransactionStatus status) {
                                 }
                             }
                         }
-                    }
-                });
-            } catch (Exception e) {
-                s_logger.warn("Error while collecting vm disk stats from hosts", e);
+                    });
+                } catch (Exception e) {
+                    s_logger.warn("Error while collecting vm disk stats from hosts", e);
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/server/StatsCollector.java:797]: we could add host details in this message

File: server/src/main/java/com/cloud/server/StatsCollector.java [hunk 2]
```diff
@@ -684,18 +684,17 @@ protected void runInContext() {
                 return;
             }
             // collect the vm disk statistics(total) from hypervisor. added by weizhou, 2013.03.
-            s_logger.trace("Running VM disk stats ...");
-            try {
-                Transaction.execute(new TransactionCallbackNoReturn() {
-                    @Override
-                    public void doInTransactionWithoutResult(TransactionStatus status) {
-                        s_logger.debug("VmDiskStatsTask is running...");
+            s_logger.debug("VmDiskStatsTask is running...");
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/server/StatsCollector.java:687]: why, this change in message and level?

File: server/src/main/java/com/cloud/server/StatsCollector.java [hunk 3]
```diff
@@ -912,10 +911,10 @@ public void doInTransactionWithoutResult(TransactionStatus status) {
                                 }
                             }
                         }
-                    }
-                });
-            } catch (Exception e) {
-                s_logger.warn("Error while collecting vm network stats from hosts", e);
+                    });
+                } catch (Exception e) {
+                    s_logger.warn("Error while collecting vm network stats from hosts", e);
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/server/StatsCollector.java:916]: again, we can now add host details to the message

**Review by DaanHoogland — APPROVED** [review]
small nitpickings :+1:

**Comment by shwstppr** [comment]
@ustcweizhou do you plan to address above comments?

@blueorangutan package

**Comment by weizhouapache** [comment]
@DaanHoogland thanks for review. modified some debug message

@shwstppr code is changed.

File: server/src/main/java/com/cloud/server/StatsCollector.java (hunk 2 above)
Comment (weizhouapache) [reply server/src/main/java/com/cloud/server/StatsCollector.java:687]: @DaanHoogland the message and level are not changed.  moved out from try block.

Removed the line which is not needed as we have debug message.
`s_logger.trace("Running VM disk stats ...");`

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
This may require testing on different hypervisors too
@blueorangutan test centos7 vmware-67u3

**Comment by DaanHoogland** [comment]
3 checksum checks that failed on vmware. not related but very disturbing!?!

**Review by DaanHoogland — APPROVED** [review]
changes are good, but the contained code is not modularised enough.

**Review by shwstppr — APPROVED** [review]
LGTM
Tested changes with a KVM env, stats working fine.

**Pull request merged**